### PR TITLE
[IPython] Support IDLE and Blender scripting module as interactive shell

### PR DESCRIPTION
**Pull request by archibate**
Related issue = #1483

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

----

### CODE REVIEW

File: python/taichi/idle_hacker.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+our_code = "__import__('taichi.idle_hacker').idle_hacker.idleipc(source)"
+
+
+def idleipc(source):
+    import os
+    import taichi as ti
+    taichi_dir = os.path.dirname(os.path.abspath(ti.__file__))
+    filename = os.path.join(taichi_dir, '.tidle_' + str(os.getpid()))
+    with open(filename, 'a') as f:
+        f.write('\n===\n' + source)
```
Comment (archibate) [inline python/taichi/idle_hacker.py:12]: We have to use this dirty method since idle and taichi is running in separate process... any idea?

**Comment by archibate** [comment]
I don't have a Blender, could you confirm this PR works for the Blender scripting module too?

**Comment by Eydcao** [comment]
> I don't have a Blender, could you confirm this PR works for the Blender scripting module too?

Hi Yubin,

My first time hearing taichi script in Blender. Does Blender have sth like IDLE for running interactive python? I could help confirm when free later.

Thx

**Comment by archibate** [comment]
Why nobody review this.. Here're tons of ad-hoc engineering effort to make the stupid IDLE happy :(

**Comment by archibate** [comment]
Is this going to be closed?

**Comment by EaryChow** [comment]
Please don't close this (╯︵╰)

**Comment by archibate** [comment]
> Please don't close this (╯︵╰)

Of course I won't! :heart: But given that people never review our PR (why??), I think we can merge on our own if they don't hit the `Request Changes` button after next release.

**Comment by yuanming-hu** [comment]
Thanks for all the contributions here! This is a useful feature and the PR should stay open. I just need more time to review given the large change (and my packed daily schedule...) Sorry about that.

File: python/taichi/idle_hacker.py [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+        path = '/usr/lib/python3.8/code.py'
+
+    print('''Hi! Dear Taichi user:
+
+  It's detected that you are using Python IDLE in **interactive mode**.
+  However, Taichi could not be fully functional due to IDLE limitation, sorry :(
+  Either run Taichi in IDLE file mode, or use IPython / Jupyter instead.
+  But we do care about your experience, no matter which shell you prefer to use.
+  So, in order to play Taichi with your favorite IDLE, we may do a dirty hack:
+  Open "{path}" and append the following line to the buttom of this file:
```
Comment (isdanni) [inline python/taichi/idle_hacker.py:54]:
```suggestion
  It's detected that you are using Python IDLE in **interactive mode**.
  However, Taichi could not be fully functional due to IDLE limitation, sorry :(
  You could either run Taichi in IDLE file mode, or use IPython / Jupyter instead.
  But we do care about your experience, no matter which shell you prefer to use.
  So, in order to play Taichi with your favorite IDLE, we may do a dirty hack:
  Open "{path}" and append the following line to the bottom of this file:
```

File: python/taichi/idle_hacker.py (hunk 1 above)
Comment (isdanni) [reply python/taichi/idle_hacker.py:26]: or we can directly wrap sockets and pass arbitrary python objects using multiprocessing? Correct me if I'm wrong, I feel like I'm missing sth here lol

Comment (archibate) [reply python/taichi/idle_hacker.py:26]: > sockets

In fact, the IDLE already using socket as IPC, and that could be an ultimate solution, however:
1. Not sure if it's invarient over different versions.
2. I'm not super familiar about sockets :(

So I'd rather write a portable file IPC instead, WDYT?

> ZeroMQ

No, Taichi is already a highly dedicated package, having too heavy dependencies could easily ban it from everyday people, sorry.

Comment (isdanni) [reply python/taichi/idle_hacker.py:26]: Ahh I see. Then I think this solution's good enough. The one that works is always the best one.

**Comment by archibate** [comment]
Hi! How about this:
1. Let me create a separate package `sourceinspect`.
1. The package is able to deal with IDLE, IPython, Jupyter, Blender, etc.
1. Move all current mocks in `taichi`, including this PR, to that external package.
1. No maintaining cost in `taichi`, we maintain the mocks in `sourceinspect` separately.
1. Include `sourceinspect` as a hard dependency of `taichi`.

WDYT about this solution? @yuanming-hu @EaryChow @isdanni Inputs are welcome!

**Pull request closed without merging**